editoast/src/views/work_schedules.rs

```diff
@@ -0,0 +1,262 @@
+    if e.message.contains(
+        r#"duplicate key value violates unique constraint "work_schedule_group_name_key""#,
+    ) {
+        WorkScheduleError::NameAlreadyUsed {
+            name: name.as_ref().to_string(),
+        }
+        .into()
+    } else {
+        e
+    }
+}
+
+#[derive(Serialize, Derivative, ToSchema)]
+#[derivative(Default)]
+struct WorkScheduleItemForm {
+    pub start_date_time: NaiveDateTime,
+    pub end_date_time: NaiveDateTime,
+    pub track_ranges: Vec<TrackRange>,
+    pub obj_id: String,
+    pub work_schedule_type: WorkScheduleType,
+}
+
+impl<'de> Deserialize<'de> for WorkScheduleItemForm {
+    fn deserialize<D>(deserializer: D) -> StdResult<WorkScheduleItemForm, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(deny_unknown_fields)]
+        struct Internal {
+            start_date_time: NaiveDateTime,
+            end_date_time: NaiveDateTime,
+            track_ranges: Vec<TrackRange>,
+            obj_id: String,
+            work_schedule_type: WorkScheduleType,
+        }
+        let internal = Internal::deserialize(deserializer)?;
+
+        // Check dates
+        if internal.start_date_time > internal.end_date_time {
+            return Err(SerdeError::custom(format!(
+                "The work_schedule start date '{}' must be before the end date '{}'",
+                internal.start_date_time, internal.end_date_time
+            )));
+        }
+
+        Ok(WorkScheduleItemForm {
+            start_date_time: internal.start_date_time,
+            end_date_time: internal.end_date_time,
+            track_ranges: internal.track_ranges,
+            obj_id: internal.obj_id,
+            work_schedule_type: internal.work_schedule_type,
+        })
+    }
+}
+
+impl WorkScheduleItemForm {
+    pub fn into_work_schedule_changeset(
+        self,
+        work_schedule_group_id: i64,
+    ) -> Result<Changeset<WorkSchedule>> {
+        let work_schedule_changeset = WorkSchedule::changeset()
+            .start_date_time(self.start_date_time)
+            .end_date_time(self.end_date_time)
+            .track_ranges(self.track_ranges)
+            .obj_id(self.obj_id)
+            .work_schedule_type(self.work_schedule_type)
+            .work_schedule_group_id(work_schedule_group_id);
+        Ok(work_schedule_changeset)
+    }
+}
+
+/// This structure is used by the post endpoint to create a work schedule
+#[derive(Serialize, Deserialize, ToSchema)]
+struct WorkScheduleCreateForm {
+    work_schedule_group_name: String,
+    work_schedules: Vec<WorkScheduleItemForm>,
+}
+
+#[utoipa::path(
+    tag = "work_schedules",
+    request_body = WorkScheduleCreateForm,
+    responses(
+        (status = 201, body = Vec<WorkSchedule>, description = "The created work schedules"),
+    )
+)]
+#[post("")]
+async fn create(
+    db_pool: Data<DbPool>,
+    data: Json<WorkScheduleCreateForm>,
+) -> Result<Json<Vec<WorkSchedule>>> {
```

Comment:
Maybe it would be more useful to simply return the work schedule group id ? It would also mean less data sent to the caller (data which the caller already knows)

Reply:
It would also allow you to add easy cleanup to your tests, by wrapping the resulting WorkScheduleGroup id in a TestFixture

Reply:
Thanks for the tip :star2: